docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,927 @@
+    - [Clean separation of concern](#clean-separation-of-concern)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+    - [Story 4](#story-4)
+    - [Story 5](#story-5)
+    - [Story 6](#story-6)
+    - [Story 7](#story-7)
+  - [High level flow](#high-level-flow)
+  - [Defining the Update Plan](#defining-the-update-plan)
+  - [MachineDeployment updates](#machinedeployment-updates)
+  - [KCP updates](#kcp-updates)
+  - [Machine updates](#machine-updates)
+  - [Infra Machine Template changes](#infra-machine-template-changes)
+  - [Remediation](#remediation)
+  - [Examples](#examples)
+    - [KCP kubernetes version update](#kcp-kubernetes-version-update)
+    - [Update worker node memory](#update-worker-node-memory)
+    - [Update worker nodes OS from Linux to Windows](#update-worker-nodes-os-from-linux-to-windows)
+  - [API Changes](#api-changes)
+    - [Machine](#machine)
+    - [MachineDeployment](#machinedeployment)
+    - [KCP](#kcp)
+    - [External Update RuntimeExtension](#external-update-runtimeextension)
+      - [`CanUpdateMachine` endpoint](#canupdatemachine-endpoint)
+      - [Request](#request)
+      - [Response](#response)
+      - [`UpdateMachine` endpoint](#updatemachine-endpoint)
+      - [Request](#request-1)
+      - [Response](#response-1)
+  - [Security Model](#security-model)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Additional Details](#additional-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+__In-place Update__: any change to a Machine spec, including the Kubernetes Version, that is performed without deleting the machines and creating a new one.
+
+__External Update Lifecycle Hook__: CAPI Lifecycle Runtime Hook to invoke external update extensions.
+
+__External Update Extension__: Runtime Extension (Implementation) is a component responsible to perform in place updates when  the `External Update Lifecycle Hook` is invoked.
+
+## Summary
+
+The proposal introduces update extensions allowing users to execute custom strategies when performing Cluster API rollouts.
+
+An External Update Extension implementing custom update strategies will report the subset of changes they know how to perform. Cluster API will orchestrate the different extensions, polling the update progress from them.
+
+If the totality of the required changes cannot be covered by the defined extensions, Cluster API will allow to fall back to the current behavior (rolling update).
+
+## Motivation
+
+Cluster API by default performs rollouts by deleting a machine and creating a new one.
```

Comment:
```suggestion
Cluster API by default performs rollouts by creating a new machine and deleting the old one.
```

Isn't the flow the other way around? This has implications for example on bare-metal when you don't have a +1 spare machine to start your rollouts with, reason why In-place updates would be needed.